### ISSUE TITLE
Performance of report filters

### ISSUE THREAD

**Issue by csordasmarton** [Ericsson/codechecker#1038]
I did some measurement on report filters. I have stored `tinyxml` with `enable-all` flag for `~850` times (589584 reports):

- **getRunResults**
```sql
-- NEW QUERY
SELECT anon_1.bug_id AS anon_1_bug_id, anon_1.checker_id AS anon_1_checker_id, anon_1.checker_message AS anon_1_checker_message,
  anon_1.severity AS anon_1_severity, anon_1.file_id AS anon_1_file_id, anon_1.status AS anon_1_status,
  anon_1.message AS anon_1_message, anon_1.author AS anon_1_author, anon_1.date AS anon_1_date, files.filename AS filename 
FROM (
  SELECT anon_2.bug_id AS bug_id, anon_2.checker_id AS checker_id, anon_2.checker_message AS checker_message,
    anon_2.severity AS severity, anon_2.file_id AS file_id, anon_2.status AS status,
    review_statuses.message AS message, review_statuses.author AS author, review_statuses.date AS date 
  FROM (
    SELECT reports.bug_id AS bug_id,
      reports.checker_id AS checker_id,
      reports.checker_message AS checker_message,
      reports.severity AS severity,
      reports.file_id AS file_id,
      max(review_statuses.status) AS status 
    FROM reports
    LEFT OUTER JOIN files ON reports.file_id = files.id
    LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
    GROUP BY reports.bug_id, reports.checker_id, reports.checker_message, reports.severity, reports.file_id
    LIMIT 500 OFFSET 0
  ) AS anon_2
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = anon_2.bug_id
) AS anon_1
LEFT OUTER JOIN files ON anon_1.file_id = files.id
ORDER BY anon_1.severity DESC
```
- **getRunReportCounts**
```sql
-- NEW QUERY
SELECT runs.id AS runs_id, runs.name AS runs_name, count(DISTINCT reports.bug_id) AS count_1 
FROM reports
LEFT OUTER JOIN files ON reports.file_id = files.id
LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
LEFT OUTER JOIN runs ON reports.run_id = runs.id
GROUP BY runs.id
```

- **getCheckerCounts**
```sql
-- NEW QUERY
SELECT anon_1.checker_id AS anon_1_checker_id, max(anon_1.severity) AS max_1, count(anon_1.bug_id) AS count_1 
FROM (
  SELECT max(reports.checker_id) AS checker_id, max(reports.severity) AS severity, reports.bug_id AS bug_id 
  FROM reports
  LEFT OUTER JOIN files ON reports.file_id = files.id
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
  GROUP BY reports.bug_id
) AS anon_1
GROUP BY anon_1.checker_id
```

- **getSeverityCounts**
```sql
-- NEW QUERY
SELECT anon_1.severity AS anon_1_severity, count(anon_1.bug_id) AS count_1 
FROM (
  SELECT max(reports.severity) AS severity, reports.bug_id AS bug_id 
  FROM reports
  LEFT OUTER JOIN files ON reports.file_id = files.id
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
  GROUP BY reports.bug_id
) AS anon_1
GROUP BY anon_1.severity
```

- **getCheckerMsgCounts**
```sql
-- NEW QUERY
SELECT anon_1.checker_message AS anon_1_checker_message, count(anon_1.bug_id) AS count_1 
FROM (
  SELECT max(reports.checker_message) AS checker_message, reports.bug_id AS bug_id 
  FROM reports
  LEFT OUTER JOIN files ON reports.file_id = files.id
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
  GROUP BY reports.bug_id
) AS anon_1
GROUP BY anon_1.checker_message
```

- **getReviewStatusCounts**
```sql
-- NEW QUERY
SELECT max(anon_1.bug_id) AS max_1, anon_1.status AS anon_1_status, count(anon_1.bug_id) AS count_1 
FROM (
  SELECT reports.bug_id AS bug_id, max(review_statuses.status) AS status 
  FROM reports
  LEFT OUTER JOIN files ON reports.file_id = files.id
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
  GROUP BY reports.bug_id
) AS anon_1
GROUP BY anon_1.status
```

- **getFileCounts**
```sql
-- NEW QUERY
SELECT files.filepath AS files_filepath, anon_1.report_count AS anon_1_report_count 
FROM files
JOIN (
  SELECT anon_2.file_id AS file_id, count(1) AS report_count 
  FROM (
    SELECT reports.bug_id AS bug_id, reports.file_id AS file_id 
    FROM reports
    LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
    GROUP BY reports.bug_id, reports.file_id
  ) AS anon_2
GROUP BY anon_2.file_id
) AS anon_1 ON anon_1.file_id = files.id
```

- **getRunHistoryTagCounts**
```sql
-- NEW QUERY
SELECT anon_1.run_id AS anon_1_run_id, runs.name AS runs_name, run_histories.time AS run_histories_time,
  run_histories.version_tag AS run_histories_version_tag, anon_2.report_count AS anon_2_report_count 
FROM (
  SELECT run_histories.run_id AS run_id, max(run_histories.id) AS run_history_id 
  FROM run_histories 
  WHERE run_histories.version_tag IS NOT NULL
  GROUP BY run_histories.run_id
) AS anon_1
LEFT OUTER JOIN run_histories ON run_histories.id = anon_1.run_history_id
LEFT OUTER JOIN runs ON runs.id = anon_1.run_id
LEFT OUTER JOIN (
  SELECT reports.run_id AS run_id, count(DISTINCT reports.bug_id) AS report_count 
  FROM reports
  LEFT OUTER JOIN files ON reports.file_id = files.id
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
  GROUP BY reports.run_id
) AS anon_2 ON anon_2.run_id = run_histories.run_id 
WHERE run_histories.version_tag IS NOT NULL
```

**Comment by csordasmarton** [comment]
`getFileCounts` function uses the following query:
```sql
-- SQL query time: 13895ms
SELECT max(reports.id) AS max_1, files.filepath AS files_filepath, count(distinct reports.bug_id) AS count_1 
FROM reports
LEFT OUTER JOIN files
  ON reports.file_id = files.id
LEFT OUTER JOIN review_statuses
  ON review_statuses.bug_hash = reports.bug_id
GROUP BY files.filepath
```

By replacing this query with the following one we can reach better performance:
```sql
-- SQL query time: 82ms
SELECT
  files.filepath,
  report_count.ct
FROM files
JOIN (
  SELECT distinct_reports.file_id, count(1) AS ct
  FROM(
    SELECT DISTINCT reports.bug_id, reports.file_id
    FROM reports
  ) as distinct_reports
  GROUP BY distinct_reports.file_id
) AS report_count 
ON report_count.file_id = files.id
order by files.filepath
```

I get the same results for both query.

**Comment by csordasmarton** [comment]
`getRunResults` with `unique` report filter:
- **Without index:**
```sql
SELECT DISTINCT reports.bug_id AS reports_bug_id, reports.checker_id AS reports_checker_id,
  reports.checker_message AS reports_checker_message, reports.severity AS reports_severity,
  files.filename AS files_filename, review_statuses.bug_hash AS review_statuses_bug_hash,
  review_statuses.status AS review_statuses_status, review_statuses.author AS review_statuses_author,
  review_statuses.message AS review_statuses_message, review_statuses.date AS review_statuses_date 
FROM reports
LEFT OUTER JOIN files
  ON reports.file_id = files.id
LEFT OUTER JOIN review_statuses
  ON review_statuses.bug_hash = reports.bug_id
ORDER BY reports.severity DESC, reports.bug_id, reports.checker_id, reports.checker_message, reports.severity, files.filename, review_statuses.status
LIMIT 500 OFFSET 0
```
**Query time:**
real	0m**17.655s**
user	0m0.301s
sys	0m0.044s

- Creating an **index on multiple column**:
```sql
--- CREATE INDEX reports_filter_column_idx ON reports (bug_id , checker_id, checker_message, severity);

SELECT anon_1.bug_id AS anon_1_bug_id, anon_1.checker_id AS anon_1_checker_id, anon_1.checker_message AS anon_1_checker_message,
  anon_1.severity AS anon_1_severity, anon_1.filename AS anon_1_filename, anon_1.status AS anon_1_status,
  review_statuses.bug_hash AS review_statuses_bug_hash, review_statuses.status AS review_statuses_status,
  review_statuses.author AS review_statuses_author, review_statuses.message AS review_statuses_message,
  review_statuses.date AS review_statuses_date
FROM (
  SELECT reports.bug_id AS bug_id, reports.checker_id AS checker_id, reports.checker_message AS checker_message,
    reports.severity AS severity, max(files.filename) AS filename, max(review_statuses.status) AS status 
  FROM reports
  LEFT OUTER JOIN files
    ON reports.file_id = files.id
  LEFT OUTER JOIN review_statuses
    ON review_statuses.bug_hash = reports.bug_id
  GROUP BY reports.bug_id, reports.checker_id, reports.checker_message, reports.severity
   LIMIT 500 OFFSET 0
) AS anon_1
LEFT OUTER JOIN review_statuses
  ON review_statuses.bug_hash = anon_1.bug_id
ORDER BY anon_1.severity DESC
```
**Query time**:
real	0m**1.136s**
user	0m0.283s
sys	0m0.039s

:exclamation: Note: if we would like to reach this speed, we have to create a `filename` column in the `reports` table and creating the following index:
```sql
CREATE INDEX reports_filter_column_idx ON reports (bug_id , checker_id, checker_message, severity, filename);
```

**Comment by Xazax-hun** [comment]
Instead of adding a filename column in the reports table could we try something else:
* Add an index to the filename column of the File table.
* Query the set of file ids based on the file name filter and run info
* Use `IN` in the query instead of a join.

**Comment by csordasmarton** [comment]
Compare old and new queries:

| API function | Old query time | New query time |
| ------------ | -------------- | -------------- |
| getRunResults | 40.2880578041 | 0.826292991638 |
| getRunReportCounts | 1.64031505585 | 1.25441908836 |
| getCheckerCounts | 26.5518729687 | 1.41373920441 |
| getSeverityCounts | 12.2330861092 | 0.177959918976 |
| getCheckerMsgCounts | 27.8827631474 | 1.61748409271 |
| getReviewStatusCounts | 12.1167159081 | 0.204226016998 |
| getFileCounts | 33.2955338955 | 0.159147024155 |
| getRunHistoryTagCounts | 8.66810107231 | 0.305584192276 |

:exclamation: Note: to reach the upper performance for `getRunResults` we have to create the following index:
```sql
CREATE INDEX reports_filter_column_idx ON reports (
  bug_id , checker_id, checker_message, severity, file_id);
```

Measurement:
- query without index: **33912ms**
- query with index: **688ms**

**Comment by csordasmarton** [comment]
We can also use the following query for `getRunResults`:
```sql
SELECT anon_1.bug_id AS anon_1_bug_id, anon_1.checker_id AS anon_1_checker_id, anon_1.checker_message AS anon_1_checker_message,
  anon_1.severity AS anon_1_severity, anon_1.file_id AS anon_1_file_id, anon_1.status AS anon_1_status,
  anon_1.message AS anon_1_message, anon_1.author AS anon_1_author, anon_1.date AS anon_1_date, files.filename AS filename 
FROM (
  SELECT anon_2.bug_id AS bug_id, anon_2.checker_id AS checker_id, anon_2.checker_message AS checker_message,
    anon_2.severity AS severity, anon_2.file_id AS file_id, anon_2.status AS status,
    review_statuses.message AS message, review_statuses.author AS author, review_statuses.date AS date 
  FROM (
    SELECT reports.bug_id AS bug_id,
      max(reports.checker_id) AS checker_id,
      max(reports.checker_message) AS checker_message,
      max(reports.severity) AS severity,
      max(reports.file_id) AS file_id,
      max(review_statuses.status) AS status 
    FROM reports
    LEFT OUTER JOIN files ON reports.file_id = files.id
    LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
    GROUP BY reports.bug_id
    LIMIT 500 OFFSET 0
  ) AS anon_2
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = anon_2.bug_id
) AS anon_1
LEFT OUTER JOIN files ON anon_1.file_id = files.id
ORDER BY anon_1.severity DESC
```
Query time: **3020ms**

:exclamation: Note:
- the **advantage** of this query is that we don't need to create indexes. 
- the **disadvantage** is that we may loss information.

**Comment by bruntib** [comment]
There are several queries which are just counting the number of certain reports. In these queries there are joined tables on files and others. I assume these joins are required because there may be a filtering condition which refers to these joined tables. Maybe it would be worth to check filter object first and skip the joins to which no filtering condition refers.

**Comment by csordasmarton** [comment]
I've tried out some queries with and without the joins and the performance was the same.

**Comment by csordasmarton** [comment]
Another solution:
Suppose we have found the same bug in different files:
`aaa` -> `tinyxml1.cpp`
`aaa` -> `tinyxml2.cpp`

It would be better to group by only by `bug_hash` and `file_id` columns and get the maximum of severity, checker id, checker message and review status columns:
```sql
SELECT anon_1.bug_id AS anon_1_bug_id, anon_1.checker_id AS anon_1_checker_id, anon_1.checker_message AS anon_1_checker_message,
  anon_1.severity AS anon_1_severity, anon_1.file_id AS anon_1_file_id, anon_1.status AS anon_1_status,
  anon_1.message AS anon_1_message, anon_1.author AS anon_1_author, anon_1.date AS anon_1_date, files.filename AS filename 
FROM (
  SELECT anon_2.bug_id AS bug_id, anon_2.checker_id AS checker_id, anon_2.checker_message AS checker_message,
    anon_2.severity AS severity, anon_2.file_id AS file_id, anon_2.status AS status,
    review_statuses.message AS message, review_statuses.author AS author, review_statuses.date AS date 
  FROM (
    SELECT reports.bug_id AS bug_id,
      max(reports.checker_id) AS checker_id,
      max(reports.checker_message) AS checker_message,
      max(reports.severity) AS severity,
      reports.file_id AS file_id,
      max(review_statuses.status) AS status 
    FROM reports
    LEFT OUTER JOIN files ON reports.file_id = files.id
    LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = reports.bug_id
    GROUP BY reports.bug_id, reports.file_id
    LIMIT 500 OFFSET 0
  ) AS anon_2
  LEFT OUTER JOIN review_statuses ON review_statuses.bug_hash = anon_2.bug_id
) AS anon_1
LEFT OUTER JOIN files ON anon_1.file_id = files.id
ORDER BY anon_1.severity DESC
```

Query time: **3605ms**
This way we don't lose so much information.

**Comment by csordasmarton** [comment]
This has been fixed in the #1121 and #1094 pull requests.
